### PR TITLE
fix bug in CRFEntityExtractor

### PR DESCRIPTION
**Pull request by shiva-z**
This PR fixes a bug in CRFEntityExtractor component. This component joins tokens with whitespaces to form the entities ([line 221](https://github.com/RasaHQ/rasa/blob/master/rasa/nlu/extractors/crf_entity_extractor.py#L221)) which is only correct if `whitespaceTokenizer` is used. For example, if we have the entity `St. Michael's Hospital`  in a message and we use `SpacyTokenizer`, the tokens would be [`St.`, `Michael`, `'s`, `Hospital`] and if we join these tokens with whitespaces we get 
`St. Michael 's Hospital`
 instead of
 `St. Michael's Hospital`

**Proposed changes**:
- Get entity values using start and end indices of tokens instead of joining tokens with whitespaces.

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by shiva-z** [comment]
> Nice find! And thanks for providing us with the fix.
> 
> Implementation is great. To prevent something like this from happening again, could you please add a test to `tests.nlu.base.test_extractors`? Covering `_create_entity_dict()` with the example you gave in your post would suffice, but if you're feeling extra helpful you can provide something more comprehensive 🙂

@MetcalfeTom  Actually I figured that https://github.com/RasaHQ/rasa/pull/4018/commits/8c8b479f24f64cbef3b141e9d98247009817e3c2 is not the best solution, here I assign a substring of the message text to entity value, but in other places in rasa, tokens are used to form entity values. Now the problem is that tokens might be **lowercased** chunks of the original message (depending on the tokenizer that is used) so here we should also use tokens to form entity values to be consistent. I fixed this issue in https://github.com/RasaHQ/rasa/pull/4018/commits/e3e0a177eca5246fc5adcd3d316f8e958fbdec32 and I will squash these two commits before merge (I thought reviewing this PR would be easier  if I leave them separate for now)

**Comment by wochinge** [comment]
@MetcalfeTom What's the state here?

File: tests/nlu/base/test_extractors.py [hunk 1]
```diff
@@ -145,6 +145,72 @@ def test_crf_json_from_non_BILOU(spacy_nlp, ner_crf_pos_feature_config):
     assert rs[4] == {"start": 29, "end": 31, "value": "by", "entity": "where"}
 
 
+def test_crf_create_entity_dict(spacy_nlp):
+    from rasa.nlu.extractors.crf_entity_extractor import CRFEntityExtractor
+    from rasa.nlu.tokenizers.spacy_tokenizer import SpacyTokenizer
+
+    crf_extractor = CRFEntityExtractor()
+    tokenizer = SpacyTokenizer()
```
Comment (MetcalfeTom) [inline tests/nlu/base/test_extractors.py]: Can you also add a `WhitespaceTokenizer` and assert that the result is the same for both?

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 2]
```diff
@@ -210,15 +210,21 @@ def most_likely_entity(self, idx, entities):
         else:
             return "", 0.0
 
-    def _create_entity_dict(self, tokens, start, end, entity, confidence):
+    def _create_entity_dict(self, message, tokens, start, end, entity, confidence):
```
Comment (MetcalfeTom) [inline rasa/nlu/extractors/crf_entity_extractor.py]: Could you add types here and a docstring?

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 3]
```diff
@@ -307,12 +313,14 @@ def _from_crf_to_json(
...
-            return self._convert_bilou_tagging_to_entity_result(tokens, entities)
+            return self._convert_bilou_tagging_to_entity_result(
+                message, tokens, entities
+            )
         else:
             # not using BILOU tagging scheme, multi-word entities are split.
             return self._convert_simple_tagging_to_entity_result(tokens, entities)
 
-    def _convert_bilou_tagging_to_entity_result(self, tokens, entities):
+    def _convert_bilou_tagging_to_entity_result(self, message, tokens, entities):
```
Comment (MetcalfeTom) [inline rasa/nlu/extractors/crf_entity_extractor.py]: Same again - types and a docstring please

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 2 above)
Comment (shiva-z) [reply rasa/nlu/extractors/crf_entity_extractor.py]: fixed here https://github.com/RasaHQ/rasa/pull/4018/commits/f65f354a4d115ce52f075aed3249740b81b96601

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 3 above)
Comment (shiva-z) [reply rasa/nlu/extractors/crf_entity_extractor.py]: fixed here https://github.com/RasaHQ/rasa/pull/4018/commits/f65f354a4d115ce52f075aed3249740b81b96601

File: tests/nlu/base/test_extractors.py (hunk 1 above)
Comment (shiva-z) [reply tests/nlu/base/test_extractors.py]: fixed here https://github.com/RasaHQ/rasa/pull/4018/commits/7427cc1935c2cb9806f27651f27a77a00ffa9747

**Comment by federicotdn** [comment]
@shiva-z Looks great!! Would you mind fixing the type checks? You can run them locally if you want; see the instructions on README.md on how to do that. After doing that and updating from `master`, we can merge this PR.

**Comment by federicotdn** [comment]
Merged! Thanks @shiva-z for your contribution!

**Pull request merged**